### ISSUE TITLE
go_test: data not available in init()

### ISSUE THREAD

**Issue by robfig** [bazelbuild/rules_go#1918]
In rules_go v0.16.5, go_test targets are not able to access data files in init() functions. I presume this is because whatever changes the working directory does not run until afterwards. However questionable reading files in init() is as a design choice, it does work with the Go command line tool, so seems like it should be supported by rules_go as well.

Here's a reproducer
https://github.com/robfig/go_test_data

**Comment by jayconrod** [comment]
Bazel starts test binaries at the root of the workspace. By default, `go_test` binaries change to the package directory before running tests. That change is done in `main`, when runs after `init` functions. This change isn't necessary with `go test`, since `cmd/go` runs test binaries in the right directories.

I can imagine a few hacks that could make this work. We could have all test packages depend on a package that changes directories in an `init`. Or we could have a wrapper that changes directories before `exec`. Error handling in both cases would be difficult though, and I'd prefer not to make this change.

Out of curiosity, are your `init` functions in libraries under test or in the test sources? If it's the latter, can you move that to `TestMain`?

**Comment by robfig** [comment]
This particular case was in the test sources, and my workaround was to wrap access to the variable that it initialized in a getter using `sync/once` to load the file. I'm happy with leaving it like that, so I don't personally consider this a high priority item to fix. 

As I make more progress on migrating to these rules, I can share different scenarios that arise.

**Comment by dragonsinth** [comment]
Oof, this leads to some very subtle, hard to diagnose bugs. :/

**Comment by dragonsinth** [comment]
> I can imagine a few hacks that could make this work. We could have all test packages depend on a package that changes directories in an `init`. Or we could have a wrapper that changes directories before `exec`. Error handling in both cases would be difficult though, and I'd prefer not to make this change.

This first idea seems tractable, at least in theory.  How does the bazel -> go_test process startup stuff actually work?  I'm having trouble following the pieces.  I know at some point there's a generated file that references the user's test package.  If that same file also referenced a package whose sole purpose was to set the working dir correctly, things would happen in the right order with a construct like:

```go
package main

import (
	_ "bazel/change_test_dir"
)

import (
	_ "user/pkg/to/test"
)

func main() {
}

```

It seems to initialize imports in lexical order within a file, and I think each file in a package is treated alphabetically.

**Comment by dragonsinth** [comment]
Aha, found it.. looking at `go/tools/builders/generate_test_main.go`.

Instead of using RunDir in the template, we'd want to link to a generic e.g. `github.com/bazelbuild/rules_go/go/tools/test_chdir`, that could maybe use `TEST_TARGET` or `TEST_BINARY` to infer the correct directory to cd to at runtime.

**Comment by jayconrod** [comment]
I think using a separate wrapper process would be the only reasonable way to make this change. If you change directories in `init`, some packages will see the initial current directory, other packages will see the package directory.

I'd strongly recommend not accessing the file system in `init` in any case because there's no error handling. Use `TestMain` instead.

**Comment by dragonsinth** [comment]
> I think using a separate wrapper process would be the only reasonable way to make this change. If you change directories in `init`, some packages will see the initial current directory, other packages will see the package directory.

You mean Go system packages?  As long as the starting package in the generated binary refers to a chdir package first, then only the chdir's own imports would ever see the original directory.  (e.g., `os`).  Which seems fine for any practical cases.

That said, a simple self-exec would be easy enough-- we wouldn't even need a separate wrapper binary.
 
> I'd strongly recommend not accessing the file system in `init` in any case because there's no error handling. Use `TestMain` instead.

I mean yeah, that's what we're having to do, practically.  It just seems like a very surprising behavior, especially for teams migrating from go test to bazel test, and it's very hard to debug initially and figure what what's going in. :/

**Comment by jayconrod** [comment]
> You mean Go system packages? As long as the starting package in the generated binary refers to a chdir package first, then only the chdir's own imports would ever see the original directory. (e.g., os). Which seems fine for any practical cases.

Maybe... seems very brittle, and only a partial solution. Constructors in linked C++ code for example would still see the initial directory.

> That said, a simple self-exec would be easy enough-- we wouldn't even need a separate wrapper binary.

You'd need a separate wrapper if the `init` code fails by panicking.

> I mean yeah, that's what we're having to do, practically. It just seems like a very surprising behavior, especially for teams migrating from go test to bazel test, and it's very hard to debug initially and figure what what's going in. :/

That's fair. It's definitely a problem.

**Comment by dragonsinth** [comment]
> > That said, a simple self-exec would be easy enough-- we wouldn't even need a separate wrapper binary.
> 
> You'd need a separate wrapper if the `init` code fails by panicking.

We could avoid ever running user init() code using the same technique we'd want to use to do an early chdir.

> > I mean yeah, that's what we're having to do, practically. It just seems like a very surprising behavior, especially for teams migrating from go test to bazel test, and it's very hard to debug initially and figure what what's going in. :/
> 
> That's fair. It's definitely a problem.

There's a couple other really nice things we could unlock here as well, for example, like setting TMPDIR = TEST_TMPDIR so that bazel go tests using os.TempDir() / io.Temp* would actually use the correct scratch directory!

**Comment by dragonsinth** [comment]
@jayconrod I'm trying an experiment with this now, but I'm kinda stuck... I created a package `github.com/bazelbuild/rules_go/go/tools/test_init` but I can't for the life of me figure out how to actually wire it into a `go_test` build correctly.  I always get
```
compilepkg: missing strict dependencies:
	/private/var/tmp/_bazel_scottb/befbf4792f1e89f7a96261537dac474e/sandbox/darwin-sandbox/6834/execroot/fullstory/bazel-out/darwin-fastbuild/bin/go/src/fs/bazel/bzimport1/bzimport1_test_/testmain.go: import of "github.com/bazelbuild/rules_go/go/tools/test_init"
No dependencies were provided.
```

I was going to try to copy along the setup for `coverdata`, but... wow... there'  a lot there, seems like way more than I'd actually need.  Can you suggest a more surgical place to add this as a compile+link dep?

**Comment by jayconrod** [comment]
`go/private/rules/test.bzl` would be the file to look at. You could add a implicit attribute to `_go_test_kwargs` depending on a `go_library` in `go/tools/builders`.

If that library needs to be generated instead, then `_go_test_impl` would be the thing to modify.

I still think a wrapper process would be better, but thinking on it more, it would be complicated to get Bazel to actually start a binary separate from the test binary. Probably not worth doing that.

**Comment by dragonsinth** [comment]
I have a proof of concept working now.

But I couldn't figure out how to wire this just into `_go_test_kwargs`... I tried setting a default value for `deps`, like so..

```bzl
"deps": attr.label_list(providers = [GoLibrary], default=["@io_bazel_rules_go//go/tools/test_init"]),
```

but it only worked for test rules that didn't have `deps` already.  I didn't see any other methods on `label_list` that would let me do anything useful?

So I ended up following all the initialization steps for `coverdata` with a new `test_init` field.  This might be overkill-- maybe there's a cleaner way but I'm pretty green with bazel internals at the skylark level.  I can send you a PR with what I did and maybe there's ways to improve it?
